### PR TITLE
Gdr 2643

### PR DESCRIPTION
**Pull request by bczech**
# Description
## What changed?
Related JIRA issue: GDR-2643

## Why was it changed?
To identify additional perturbations hidden in the data

# Checklist for sustainable code base
- [X] I added tests for any code changed/added
- [X] I added documentation for any code changed/added
- [X] I made sure naming of any new functions is self-explanatory and consistent

# Logistic checklist
- [X] Package version bumped
- [X] Changelog updated

# Screenshots (optional)

### CODE REVIEW

File: R/data_type.R [hunk 1]
```diff
@@ -324,3 +341,82 @@ collapse_drugs <- function(df) {
...
+      if (length(unique_drugs) == 1) {
+        # Check if there are only two doses in the current concentration column (0 and another value)
+        unique_doses <- unique(dt[[conc_col]])
+        
+        if (length(unique_doses) == 2 && 0 %in% unique_doses) {
+          # Create a new column named after the drug (excluding untreated_tag) and fill it with the doses
+          new_column_name <- unique_drugs[1]
+          dt[, (new_column_name) := dt[[conc_col]]]
+          
+          drug_order <- gsub(".*_(\\d+)$|.*", "\\1", drug_col)
```
Comment (gladkia) [inline R/data_type.R:409]: Add a test to ensure that this chunks works as expected.

Comment (bczech) [reply R/data_type.R:409]: done

**Review by gladkia — APPROVED** [review]

File: R/data_type.R [hunk 2]
```diff
@@ -44,25 +44,30 @@ identify_data_type <- function(df,
                                codilution_conc = 2,
                                matrix_conc = 1) {
   
-  # find the pairs of drugs with relevant metadata
```
Comment (j-smola) [inline R/data_type.R:47]: Maybe that would be a god idea to rename `df` in something more descriptive? `dt_response`?

Comment (bczech) [reply R/data_type.R:47]: renamed into dt

File: R/data_type.R [hunk 3]
```diff
@@ -44,25 +44,30 @@ identify_data_type <- function(df,
...
-  ))
-  conc_ids <- conc_ids[which(conc_ids %in% names(df))]
+  # Get drug and concentration identifiers
+  drug_ids <- get_relevant_ids(c("drug_name", "drug_name2", "drug_name3"), df)
+  conc_ids <- get_relevant_ids(c("concentration", "concentration2", "concentration3"), df)
   
-  drugs_cotrt_ids <- drugs_ids[!names(drugs_ids) %in% "drug_name"]
-  conc_cotrt_ids <- conc_ids[!names(conc_ids) %in% "concentration"]
+  # Filter out primary drug and concentration identifiers
+  drugs_cotrt_ids <- setdiff(drug_ids, "drug_name")
```
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R:52]: thanks, done

File: R/data_type.R [hunk 4]
```diff
@@ -252,6 +257,18 @@ split_raw_data <- function(df,
   }, df_list)
 }
 
+#' @keywords internal
+#' Function to get relevant identifiers from the environment
+#' @param identifiers A character vector of identifier names to fetch from the environment
+#' @param df A data frame containing the columns to be checked against the identifiers
```
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R:263]: fixed

File: R/data_type.R [hunk 5]
```diff
@@ -252,6 +257,18 @@ split_raw_data <- function(df,
...
+#' @return A character vector of relevant identifiers that are present in the data frame
+#' @keywords internal
+get_relevant_ids <- function(identifiers, df) {
```
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R:266]: done

File: R/data_type.R [hunk 6]
```diff
@@ -324,3 +341,84 @@ collapse_drugs <- function(df) {
...
+#' it creates a new column named after the drug and fills it with the doses,
+#' then removes the original drug and concentration columns.
+#'
+#' @param dt A data.table containing the data.
+#' @param drugs_cotrt_ids A vector of column names related to drugs.
+#' @param conc_cotrt_ids A vector of column names related to concentrations.
+#' @param untreated_tag A string representing the untreated tag (default is "vehicle").
+#' @return A modified data.table with new columns for the drugs and removed original drug and concentration columns.
+#' @examples
+#' library(data.table)
```
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R:362]: done

File: R/data_type.R [hunk 7]
```diff
@@ -324,3 +341,84 @@ collapse_drugs <- function(df) {
...
+                                  drugs_cotrt_ids,
+                                  conc_cotrt_ids,
+                                  untreated_tag = "vehicle") {
+  
+  # Assertions
+  checkmate::assert_data_table(dt)
+  checkmate::assert_character(drugs_cotrt_ids, any.missing = FALSE)
+  checkmate::assert_character(conc_cotrt_ids, any.missing = FALSE)
+  checkmate::assert_true(length(drugs_cotrt_ids) == length(conc_cotrt_ids))
+
```
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R:384]: used another assert

File: R/data_type.R [hunk 8]
```diff
@@ -324,3 +341,84 @@ collapse_drugs <- function(df) {
...
+#' print(dt)
+#' @export
+process_perturbations <- function(dt,
+                                  drugs_cotrt_ids,
+                                  conc_cotrt_ids,
+                                  untreated_tag = "vehicle") {
+  
+  # Assertions
+  checkmate::assert_data_table(dt)
+  checkmate::assert_character(drugs_cotrt_ids, any.missing = FALSE)
```
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R:381]: done

File: R/data_type.R [hunk 9]
```diff
@@ -44,25 +44,30 @@ identify_data_type <- function(df,
...
+  # Get untreated tag and cell line identifiers
   untreated_tag <- c(gDRutils::get_env_identifiers("untreated_tag"), NA)
   cell <- gDRutils::get_env_identifiers("cellline_name")
-  cols_pairs <- intersect(names(df),  c(drug_ids, cell, conc_ids))
+  
+  # Process perturbations
+  dt <- process_perturbations(df, drugs_cotrt_ids, conc_cotrt_ids, untreated_tag)
+  
+  # Double-check the columns after processing perturbations
+  drug_ids <- get_relevant_ids(c("drug_name", "drug_name2", "drug_name3"), df)
```
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R:63]: done

File: R/data_type.R [hunk 10]
```diff
@@ -324,3 +341,84 @@ collapse_drugs <- function(df) {
...
+  checkmate::assert_data_table(dt)
+  checkmate::assert_character(drugs_cotrt_ids, any.missing = FALSE)
+  checkmate::assert_character(conc_cotrt_ids, any.missing = FALSE)
+  checkmate::assert_true(length(drugs_cotrt_ids) == length(conc_cotrt_ids))
+  
+  
+  # If lengths of drugs_cotrt_ids and conc_cotrt_ids are 0, return dt unchanged
+  if (length(drugs_cotrt_ids) == 0 && length(conc_cotrt_ids) == 0) {
+    return(dt)
+  } else {
```
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R]: done

File: R/data_type.R [hunk 11]
```diff
@@ -233,32 +238,46 @@ split_raw_data <- function(df,
...
     selected_columns <- which(names(x) != type_col)
     x[, selected_columns, with = FALSE]
-  }, df_list)
+  }, dt_list)
+}
+
+#' @keywords internal
+#' Function to get relevant identifiers from the environment
+#' @param identifiers A character vector of identifier names to fetch from the environment
+#' @param dt A data frame containing the columns to be checked against the identifiers
```
Comment (j-smola) [inline R/data_type.R:263]:
```suggestion
#' @param dt data.table containing the columns to be checked against the identifiers
```

File: R/data_type.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R]: ✔️

File: R/data_type.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R]: ✔️

File: R/data_type.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R]: ✔️

File: R/data_type.R (hunk 10 above)
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R]: ✔️

File: R/data_type.R [hunk 12]
```diff
@@ -324,3 +341,84 @@ collapse_drugs <- function(df) {
...
+          new_column_name <- unique_drugs[1]
+          dt[, (new_column_name) := dt[[conc_col]]]
+          
+          drug_order <- gsub(".*_(\\d+)$|.*", "\\1", drug_col)
+          drug_cols <- 
+            unique(c(
+              get_relevant_ids(paste0(
+              c("drug", "drug_name", "drug_moa", "concentration"), drug_order),
+              dt),
+              drug_col, conc_col))
```
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R:421]: ✔️

File: R/data_type.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R:392]: ✔️

File: R/data_type.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R]: ✔️

File: R/data_type.R (hunk 9 above)
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R]: ✔️

File: R/data_type.R (hunk 2 above)
Comment (j-smola) [reply R/data_type.R:47]: ✔️

File: R/data_type.R (hunk 8 above)
[the remark this answers is not in the document]
Comment (j-smola) [reply R/data_type.R]: So, the assumption is that first value from `drugs_cotrt_ids` corresponds to the first value from `conc_cotrt_ids`. And function check this pair.

@bczech But there is one more minor problem. The `process_perturbations` function modifies the source object. In the above example, `dt3` after processing by `process_perturbations` will be as the `result`. Maybe this side effect could be removed?

File: R/data_type.R [hunk 13]
```diff
@@ -303,24 +322,109 @@ collapse_drugs <- function(df) {
...
+#'   drug1 = c("vehicle", "drugA", "drugA"),
+#'   conc1 = c(0, 10, 0),
+#'   drug2 = c("vehicle", "drugB", "drugB"),
+#'   conc2 = c(0, 20, 0)
+#' )
+#' drugs_cotrt_ids <- c("drug1", "drug2")
+#' conc_cotrt_ids <- c("conc1", "conc2")
+#' dt <- process_perturbations(dt, drugs_cotrt_ids, conc_cotrt_ids)
+#' print(dt)
+#' @export
```
Comment (j-smola) [inline R/data_type.R:374]: add missing `@keywords`

File: R/data_type.R (hunk 11 above)
Comment (bczech) [reply R/data_type.R:263]: done

File: R/data_type.R (hunk 8 above)
[the remark this answers is not in the document]
Comment (bczech) [reply R/data_type.R]: done

File: R/data_type.R [hunk 14]
```diff
@@ -390,8 +390,10 @@ process_perturbations <- function(dt,
   checkmate::assert_true(length(drugs_cotrt_ids) == length(conc_cotrt_ids))
   checkmate::assert_subset(untreated_tag, c(gDRutils::get_env_identifiers("untreated_tag"), NA))
   
+  # Create a copy of dt to avoid modifying the input in place
+  dt_copy <- data.table::copy(dt)
```
Comment (j-smola) [inline R/data_type.R:394]: 👍🏻

**Review by j-smola — APPROVED** [review]

**Pull request merged**